accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -247,10 +250,10 @@ not a subtype of the class type associated with the class in which it is defined
 constructor for any class other than `Null`).
 
 It is a warning to use a null aware operator (`?.`, `?..`, `??`, `??=`, or
-`...?`) on a non-nullable value.
+`...?`) on a value of type `T` if `T <: Object` and not `Null <: T`.
```

Comment:
There is a syntactic criterion for this, which is exactly the same as the syntactic criterion for non-nullable, except with this line deleted:

  - `S*` for some `S` where `S` is non-nullable

I will update this tomorrow to use that.

Reply:
I want `?.` (etc.) to be allowed for definitely nullable types (`Null` <: T), legacy types (`Null` <: T && T <: `Object`), `dynamic`, and potentially nullable type variables (`X` where `X extends T?`). Basically, for anything which *can* be `null`. 
The type variable is a tricky case because it does not satisfy `Null` <: X.

That is, it's only an error if the type *definitely cannot* contain `null`, so a phrasing like "unless T is potentially nullable" sounds exactly right. We have already defined that anyway.

Reply:
Using potentially nullable is just wrong.  Using this criteria:
```
      a?.b is an error unless Typeof(a) is potentially nullable
```
We have, if `TypeOf(a)` is `int*`, that 
  - `not(potentially nullable int*) => error`
  - `not(not (non-nullable int*)) => error`
  - `non-nullable int* => error`
  - `true => error`

The criteria I gave on the other hand gives exactly what you want:
  - non legacy nullable types are allowed (`Null <: T`)
  - legacy types are allowed (`Null <: T`)
  - dynamic is allowed (`Null <: T`)
  - Type variables `X <: Object?` are allowed (not `T <: Object`)
  - Type variables `X <: Object*` are allowed (`T <: Null`)
  - Type variables `X <: Object` are not allowed (`T <: Object`)
  - All soundly non-nullable types are disallowed (`T <: Object` and not `Null <: T`).

Reply:
I added a syntactic criteria for this.